File: projects/hanoi_learn_swift/Sources/Tour/objects_and_classes.swift
func objects_and_classes() {

    // simple class
    class EPower0 {
        var U: Double
        var I: Double
        var brand: String
        var year: Int
        var leap: Bool

        init(U: Double, I: Double) {

            self.U = U
            self.I = I

            self.brand = "GE"
            self.year = 1995

            let div = 1995 % 4
            self.leap = div == 0
        }

        func getWatt() -> Double {
            return U * I
        }
    }

    let EP0: EPower0 = EPower0(U: 7, I: 2)
    EP0.brand = "NVIDIA"

    print(EP0.brand)
    print(EP0.getWatt())

    class EPower1: EPower0 {
        var tz: String

        override init(U: Double, I: Double) {

            self.tz = "GMT+7"
            super.init(U: U, I: I)
        }

        func getResistance() -> Double {
            return U / I
        }
    }

    let EP1 = EPower1(U: 7, I: 2)

    print(EP1.brand)
    print(EP1.getWatt())
    print(EP1.getResistance())

    class HasStaticVar {
        static var name = "(1)"
    }

    let HSV = HasStaticVar()

    // - via class
    // Access via class
    print(HasStaticVar.name)  // (1)

    // Modify via class
    HasStaticVar.name += "(2)"
    print(HasStaticVar.name)  // "(1)(2)"

    class Person {

        private var _name: String

        init(name: String) {
            self._name = name
        }

        var name: String {

            // getter
            get {
                return _name
            }

            // setter
            set {
                _name = newValue
            }
        }
    }

    let person = Person(name: "Alice")
    print(person.name)

    person.name = "Bob"
    print(person.name)

}

File: projects/hanoi_learn_python/src/hanoi_learn_python/tour/objects_and_classes.py
#!/usr/bin/env python3



def objects_and_classes():
    class EPower0:

        def __init__(self, U, I):
            self.U = U
            self.I = I

            self.brand = 'GE'
            self.year = 1995



        def get_watt(self):
            return self.U * self.I

    EP0 = EPower0(7, 2)
    EP0.brand = 'NVIDIA'

    print(EP0.brand)
    print(EP0.get_watt())



    class EPower1(EPower0):

        def __init__(self, U, I):
            super().__init__(U, I)
            self.tz = 'GMT+7'



        def get_resistance(self):
            return self.U / self.I

    EP1 = EPower1(7, 2)
    print(EP1.brand)
    print(EP1.get_watt())
    print(EP1.get_resistance())
    # fmt: on



    class HasStaticVar:
        name = '(1)'

    HSV = HasStaticVar()
    # - via instance
    # access via instance
    print(HSV.name)  # (1)

    # modify via instance
    HSV.name += '(2)'
    print(HSV.name)  # (1)(2)

    # - via class
    # access via class
    print(HasStaticVar.name)  # (1)

    # modify via class
    HasStaticVar.name += '(3)'
    print(HasStaticVar.name)  # (1)(3)
    print(HasStaticVar().name)  # (1)(3)

    # Dont use work-around if language doesnt support simple way to modify / access static var from class / instance

    # ---- getter setter
    # fmt: off
    class Person:

        def __init__(self, name):
            self._name = name

        # getter
        @property
        def name(self):
            return self._name

        # setter
        @name.setter
        def name(self, value):
            self._name = value
    # fmt: on

    person = Person("Alice")
    print(person.name)

    person.name = "Bob"
    print(person.name)
